src/main/java/org/apache/mesos/specification/DefaultTaskSpecification.java

```diff
@@ -45,26 +46,102 @@ public static DefaultTaskSpecification create(Protos.TaskInfo taskInfo, Optional
         } catch (InvalidProtocolBufferException e) {
             throw new InvalidTaskSpecificationException("Failed to deserialize config files: " + e.getMessage());
         }
-        return new DefaultTaskSpecification(
-                taskInfo.getName(),
-                taskInfo.getCommand(),
-                getResources(taskInfo),
-                getVolumes(taskInfo),
-                configFiles,
-                placement,
-                Optional.empty());
+
+        if (!taskInfo.hasCommand() && !taskInfo.hasContainer()) {
+            throw new InvalidTaskSpecificationException("Both ContainerInfo and CommandInfo are missing, specify one.");
+        }
+
+        Optional<Protos.HealthCheck> healthCheck = taskInfo.hasHealthCheck() ?
+                Optional.of(taskInfo.getHealthCheck()) : Optional.empty();
+
+        if (taskInfo.hasCommand() && taskInfo.hasContainer()) {
+            return new DefaultTaskSpecification(
+                    taskInfo.getName(),
+                    taskInfo.getContainer(),
+                    taskInfo.getCommand(),
+                    getResources(taskInfo),
+                    getVolumes(taskInfo),
+                    configFiles,
+                    placement,
+                    healthCheck);
+        } else if (taskInfo.hasCommand()) {
+            return new DefaultTaskSpecification(
+                    taskInfo.getName(),
+                    taskInfo.getCommand(),
+                    getResources(taskInfo),
+                    getVolumes(taskInfo),
+                    configFiles,
+                    placement,
+                    healthCheck);
+        } else {
+            return new DefaultTaskSpecification(
+                    taskInfo.getName(),
+                    taskInfo.getContainer(),
+                    getResources(taskInfo),
+                    getVolumes(taskInfo),
+                    configFiles,
+                    placement,
+                    healthCheck);
+        }
+    }
+
+    /**
+     * Create task with only {@link Protos.CommandInfo}.
+     */
+    protected DefaultTaskSpecification(
+            String name,
```

Comment:
I'll fix this by making all the constructors now go via a private constructor. That removes the bad duplication. Thanks!